### DISCUSSION TITLE
Allow non-core code to write to the "History" (Audit log)

### DISCUSSION THREAD

**Discussion by hfloyd** [umbraco/Umbraco-CMS#16771] [Features and ideas]
I bumped up against a need for this when working with Umbraco Deploy (see: https://github.com/umbraco/Umbraco.Deploy.Issues/issues/22) and the answer was that the Audit log is rather limited, in terms of access. 

I think it would be great if it were possible to programmatically insert things into a node's Audit log. Right now the log shows core operations (Move, Sort, Save, Publish), but since there is so much custom functionality being added into Umbraco, and especially stuff automatically running on node events/notifications, or via some external updating routines, it would be nice if developers could add custom audit info. 

For example, say you have code that runs on a schedule to programmatically update some nodes based on an import or API call. Being able to not only set the new property values, and Save & Publish, but to add "Updated via Product Import" to the History tab, with a timestamp would be helpful for Editors and Devs troubleshooting things.

In addition, Umbraco Deploy could then insert useful information like "Heather - Deployed to Live", "Joe - Restored from Staging",  on nodes, which is a lot more useful than just "Administrator - Save & Publish".

**Comment by ronaldbarendse** [comment]
The CMS already has the `IAuditService` that allows programmatically adding entries to the history/audit log:

```c#
await _auditService.AddAsync(AuditType.Custom, userId, entityId, entityType, null, "This is my custom audit message");
```

<img width="647" height="171" alt="image" src="https://github.com/user-attachments/assets/798534f1-93eb-4048-8d74-210b9e56277b" />

The `AuditType` can't be extended though and the message needs to be passed in the last `parameters` parameter, instead of the `comment` one to actually show up in the backoffice UI. Also don't confuse this with the `IAuditEntryService`, which is a more low-level audit log of all user actions (like logins, permission and role changes, etc.).

**Reply by hfloyd** [reply]
Thanks for the tips. You can add **"ability to extend `AuditType`"** to the wishlist, then 😉

**Comment by bjarnef** [comment]
It would be useful if we can benefits from `IAuditService` and extension point for backoffice UI.
E.g. something similar to activity logs I used in Umbraco Commerce (formerly Vendr).
https://our.umbraco.com/packages/website-utilities/vendr-reviews/ 

It seems Umbraco Commerce use audits now, but not sure if its a custom implementation:
https://github.com/umbraco/Umbraco.Commerce.Issues/issues/37
